test/oban/job_test.exs

```diff
@@ -56,16 +56,37 @@ defmodule Oban.JobTest do
         |> Map.keys()
       end
 
-      assert to_keys.([]) == [:args, :queue, :state]
-      assert to_keys.(worker: MyWorker) == [:args, :queue, :state, :worker]
+      keys_with_defaults = ~w(args attempt errors max_attempts queue state)a
 
-      assert to_keys.(schedule_in: 1, worker: MyWorker) == [
-               :args,
-               :queue,
-               :scheduled_at,
-               :state,
-               :worker
-             ]
+      default_keys_with = fn extra_keys ->
+        (keys_with_defaults ++ extra_keys) |> Enum.sort() |> Enum.uniq()
+      end
+
+      assert to_keys.([]) == keys_with_defaults
+      assert to_keys.(worker: MyWorker) == default_keys_with.([:worker])
+
+      assert to_keys.(schedule_in: 1, worker: MyWorker) ==
+               default_keys_with.([:scheduled_at, :worker])
+    end
+
+    test "retains all non-nil permitted attributes" do
+      attrs = [
+        attempt: 1,
+        attempted_by: ["foo"],
+        attempted_at: DateTime.utc_now(),
+        completed_at: DateTime.utc_now(),
+        errors: [%{err: "baz"}],
+        inserted_at: DateTime.utc_now(),
+        max_attempts: 5,
+        queue: "default",
+        scheduled_at: DateTime.utc_now(),
+        state: "scheduled",
+        worker: "Fake"
+      ]
+
+      map = %{foo: "bar"} |> Job.new(attrs) |> Job.to_map()
+      assert map[:args] == %{foo: "bar"}
+      attrs |> Enum.each(fn {k, v} -> assert map[k] == v end)
```

Comment:
Extremely thorough, love it. A couple minor nitpicks:

1. Please avoid piping a single argument
2. Please avoid single character variables

```suggestion
      Enum.each(attrs, fn {key, val} -> assert map[key] == val end)
```